[precompile] Skip serializing unnecesssary objects for guards.

zhxchen17 opened the pull request:
Summary:
The following type of objects don't need to be serialized for precompile:
1. PyCapsule because we don't guard on C binding objects in meaningful ways.
2. Code object because we only id matching on these but id matches will always be dropped for precompile.
3. Nested function objects since we also ban CLOSURE_MATCH.

Test Plan:
buck run mode/opt test/dynamo:test_dynamo -- -k test_skipped_objects

Rollback Plan:

Differential Revision: D78816888




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

jamesjwu commented on torch/_dynamo/guards.py:
```diff
@@ -2792,6 +2796,22 @@ def reducer_override(self, obj):
 
+        elif (
+            obj.__class__.__module__ == "builtins"
+            and obj.__class__.__name__ == "PyCapsule"
```
Can this not be `isinstance(obj, builtins.PyCapsule)`?

zhxchen17 commented:
@pytorchbot merge

zhxchen17 committed:
[precompile] Skip serializing unnecesssary objects for guards. (#158926)

Summary:

The following type of objects don't need to be serialized for precompile:
1. PyCapsule because we don't guard on C binding objects in meaningful ways.
2. Code object because we only id matching on these but id matches will always be dropped for precompile.
3. Nested function objects since we also ban CLOSURE_MATCH.

Test Plan:
buck run mode/opt test/dynamo:test_dynamo -- -k test_skipped_objects

Rollback Plan:

Reviewed By: jamesjwu

Differential Revision: D78816888
